packages/platforms/accton/arm64/as456x-52p/src/modules/arm64-accton-as456x-52p-mux.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (C)  Brandon Chuang <brandon_chuang@accton.com.tw>
+ *
+ * This module supports the accton cpld that hold the channel select
+ * mechanism for other i2c slave devices, such as SFP.
+ * This includes the:
+ *	 Accton as456x CPLD1/CPLD2/CPLD3
+ *
+ * Based on:
+ *	pca954x.c from Kumar Gala <galak@kernel.crashing.org>
+ * Copyright (C) 2006
+ *
+ * Based on:
+ *	pca954x.c from Ken Harrenstien
+ * Copyright (C) 2004 Google, Inc. (Ken Harrenstien)
+ *
+ * Based on:
+ *	i2c-virtual_cb.c from Brian Kuschak <bkuschak@yahoo.com>
+ * and
+ *	pca9540.c from Jean Delvare <khali@linux-fr.org>.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/slab.h>
+#include <linux/device.h>
+#include <linux/i2c.h>
+#include <linux/i2c-mux.h>
+#include <linux/version.h>
+#include <linux/stat.h>
+#include <linux/delay.h>
+#include <dt-bindings/mux/mux.h>
+
+#define DRVNAME "as456x_mux"
+
+#define I2C_RW_RETRY_COUNT 10
+#define I2C_RW_RETRY_INTERVAL 60 /* ms */
+
+#define CPLD_MAX_NCHANS 6
+#define CPLD_CHANNEL_SELECT_REG 0x0
+#define CPLD_CHANNEL_DESELECT_VAL 0x0
+
+#define PCA9641_DESELECT_CHANNEL_REG 0x1
+#define PCA9641_DESELECT_CHANNEL_VAL 0x0
+
+enum mux_type {
+	pca9641_mux,
+	as456x_cpldm_mux
+};
+
+struct chip_desc {
+	u8 nchans;
+	u8 deselectChan;
+};
+
+struct as456x_mux_data {
+	enum mux_type type;
+	struct mutex update_lock;
+	struct i2c_client *client;
+};
+
+/* Provide specs for the as456x CPLD types we know about */
+static const struct chip_desc chips[] = {
+	[pca9641_mux] = {
+		.nchans = 1,
+		.deselectChan = PCA9641_DESELECT_CHANNEL_VAL
+	},
+	[as456x_cpldm_mux] = {
+		.nchans = 6,
+		.deselectChan = CPLD_CHANNEL_DESELECT_VAL
+	}
+};
+
+static const struct i2c_device_id as456x_mux_id[] = {
+	{ "pca9641_mux", pca9641_mux },
+	{ "as456x_cpldm_mux", as456x_cpldm_mux },
+	{ }
+};
+MODULE_DEVICE_TABLE(i2c, as456x_mux_id);
+
+static const struct of_device_id as456x_mux_of_match[] = {
+	{ .compatible = "edgecore,pca9641_mux", .data = &chips[pca9641_mux] },
```

Comment:
Is there anything special about this PCA9641, or could this be replaced by importing e.g. https://patchwork.ozlabs.org/project/linux-i2c/patch/20190306231521.29367-6-peda@axentia.se/#2184718 into our kernel tree (with a bit of clean-up/forward porting)?

If the latter, I would heavily suggest doing so instead, and maybe even pick up the work of submitting the support again (not necessarily by you).

Reply:
If it matters, the series that you linked seems to time out the i2c controller.  I think it might have something to do with the way this mux is being used (there are two controllers attached but only one is actually powered at a time) - I didn't dive too deeply but it looks like the chip doesn't arbitrate correctly if there aren't two controllers on it.  Directly programming the mux like this driver doesn't have the same effect.  More investigation is probably needed to be sure but there seems to be reasonable technical reason not to use the other driver.

Reply:
Thank you for testing this. Could you please also relay this test result to the LKML?